### ISSUE TITLE
Comprehensions do not split as far as they can go, which preserves long line lengths

### ISSUE THREAD

**Issue by rogersheu** [astral-sh/ruff#12870]
**Keywords**
Line length, comprehension

**Example**
*Before `ruff --fix`
```
really_really_really_long_dict_name = {"foo": "bar"}

dict_with_really_long_names = [(really_really_really_long_key_name, really_really_really_long_key_value) for really_really_really_long_key_name, really_really_really_long_key_value in really_really_really_long_dict_name.items()]
```

After `ruff --fix`

```
dict_with_really_long_names = [
    (really_really_really_long_key_name, really_really_really_long_key_value)
    for really_really_really_long_key_name, really_really_really_long_key_value in really_really_really_long_dict_name.items()
]
```

That `for/in` line is 127 characters long, but it could come under the line length (default of 88, not set here) by splitting the `for/in` statements, which does not affect behavior in this case.

**Version/Command**
Using ruff v0.5.1 with pre-commit calling `ruff --fix`.

**Comment by MichaReiser** [comment]
To clarify. Are you running `ruff format` or `ruff check --fix`? I  suspect it's the former because `E501` doesn't come with an autofix.

**Comment by MichaReiser** [comment]
Thanks.  Yeah, it is then the formatter that doesn't split the comprehension further. 

The main question here is what should happen when the comprehension does break over multiple lines:

```diff
 lcomp3 = [
         # This one is actually too long to fit in a single line.
         element.split("\n", 1)[0]
         yup
-        for element in collection.select_elements()
+        for element
+        in collection.select_elements()
         # right
         if element is not None
 ]
```

Because this does look worse. The desired result is probably that the `for ... in` should be kept together if possible

**Comment by MichaReiser** [comment]
I think my biggest concern is that it hurts readability if there are multiple generators

```python
[
    [
        x 
        for x
        in bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb
        for y 
        in xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxdxxxxxxxxxxxxxxxxxxxxxxxx
    ]
]
```

**Comment by MichaReiser** [comment]
Thanks for linking to the relevant black issue. This is very helpful for both projects :)

**Comment by MichaReiser** [comment]
There are surprisingly few differences in the [ecosystem check of a naive implementation](https://github.com/astral-sh/ruff/pull/12908#issuecomment-2291695892).

I'm not convinced that this looks better overall. It would require more refined rules. E.g. don't split if the left side is a simple expression like an identifier.

**Comment by rogersheu** [comment]
Thanks for giving this issue some thought @MichaReiser !

> The desired result is probably that the for ... in should be kept together if possible

I agree. If the `for ... in` can be kept to one line, it absolutely should. I'm purely thinking of cases where doing so would improve the soft adherence to the line length setting.

> There are surprisingly few differences in the https://github.com/astral-sh/ruff/pull/12908#issuecomment-2291695892.

> I'm not convinced that this looks better overall. It would require more refined rules. E.g. don't split if the left side is a simple expression like an identifier.

Interesting, thanks for doing that prototyping! Yeah, I assume most folks aren't using unnecessarily long variable or iterable variable names like I am (likely shoddy choices on my part), but the cases your implementation covers do seem to cover subset where the code has parentheses in the `for (...) in ...` and the parentheses get split out, perhaps unnecessarily.

For the prefect one, I think ideally, it could be the following, where trying to keep the `in` with the `for` led to the parentheses splitting out line-wise. (So the change was effective, but perhaps it could compress the earlier lines.) That might start needing some extra logic though, and I lack the intuition in this codebase to know if that's feasible and even stylistically reasonable.

```
            for (block_schema, _, parent_block_schema_id) 
            in block_schemas_with_references
```

**Comment by MichaReiser** [comment]
> likely shoddy choices on my part), but the cases your implementation covers do seem to cover subset where the code has parentheses in the for (...) in ... and the parentheses get split out, perhaps unnecessarily.

I think the reason why the parentheses get split out is because the tuple contains a trailing comma. But I think the style could do better by trying to keep the closing `)` and the `in` keyword on the same line.

**Comment by spaceone** [comment]
https://github.com/astral-sh/ruff/issues/19231 was a duplicate for this:

I am generally allowing a lower line length: 120 chars because the code base has function calls with many parameters.
On the other side I format certain list-comprehension code blocks for readability into multiple lines.

one example:
```diff
-    return [
-        idx * 8 + bit
-        for idx, value in enumerate(octets)
-        for bit in range(8)
-        if value & (1 << bit)
-    ]
+    return [idx * 8 + bit for idx, value in enumerate(octets) for bit in range(8) if value & (1 << bit)]
```

It would be nice if the formatter would allow a setting, that already multiline-splitted list/set/dict comprehensions are not joined into one single line again.
Maybe a line-length for comprehensions may differ from a general line length for function calls / the rest.

Basically that is the number one blocking issue for using the formatter at all.

**Comment by MichaReiser** [comment]
This is also related to https://github.com/astral-sh/ruff/issues/11753
